frontend_tests/puppeteer_tests/07-navigation.js

```diff
@@ -0,0 +1,106 @@
+"use strict";
+
+const assert = require("assert").strict;
+
+const common = require("../puppeteer_lib/common");
+
+async function wait_for_tab(page, tab) {
+    const tab_slector = `#${tab}.tab-pane.active`;
+    await page.waitForSelector(tab_slector, {visible: true});
+}
+
+async function navigate_to(page, click_target, tab) {
+    console.log("Visiting #" + click_target);
+    await page.click(`a[href='#${click_target}']`);
+
+    await wait_for_tab(page, tab);
+}
+
+async function open_menu(page) {
+    const menu_selector = "#settings-dropdown";
+    await page.waitForSelector(menu_selector, {visible: true});
+    await page.click(menu_selector);
+}
+
+async function navigate_to_settings(page) {
+    console.log("Navigating to settings");
+
+    await open_menu(page);
+
+    const settings_selector = "a[href^='#settings']";
+    await page.waitForSelector(settings_selector, {visible: true});
+    await page.click(settings_selector);
+
+    await page.waitForSelector("#settings_page", {visible: true});
+
+    await page.click("#settings_page .content-wrapper .exit");
+}
+
+async function navigate_to_subscriptions(page) {
+    console.log("Navigate to subscriptions");
+
+    await open_menu(page);
+
+    const manage_streams_selector = 'a[href^="#streams"]';
+    await page.waitForSelector(manage_streams_selector, {visible: true});
+    await page.click(manage_streams_selector);
+
+    await page.waitForSelector("#subscription_overlay", {visible: true});
+    await page.waitForSelector("#subscriptions_table", {visible: true});
+
+    await page.click("#subscription_overlay .exit");
+}
+
+async function test_reload_hash(page) {
+    const initial_page_load_time = await page.evaluate(() => page_params.page_load_time);
+    console.log("initial load time: " + initial_page_load_time);
+
+    const initial_hash = await page.evaluate(() => window.location.hash);
+
+    await Promise.all([
+        page.evaluate(() => reload.initiate({immediate: true})),
+
+        // navigations during reload were causing failures,
+        // this fixes it.
+        page.waitForNavigation({waitUntil: "networkidle0"}),
```

Comment:
Can you expand on this comment?  

It seems unlikely to me that waiting for idle network is precisely the right thing to wait for.

Reply:
That doesn't seem to be the right fix. It was giving `Error: Execution context was destroyed, most likely because of a navigation` due to the reload. The right fix seems to be `waitForSelector("#zhome", {visible: true});` but this seems to give a flake that I didn't see previously. It is failing with the `assert` just below this. (`assert(page_load_time > initial_page_load_time)`. Not sure what the reason is. [Here](https://circleci.com/gh/chdinesh1089/zulip/1594#artifacts/containers/0) is the failing build. Trying to debug that now.

Reply:
waiting for `#zhome` wasn't right as it doesn't appear in a stream narrow. Change to `#zfilt` and a 100 run build [passed](https://circleci.com/gh/chdinesh1089/zulip/1606) Also added [another](https://circleci.com/gh/chdinesh1089/zulip/1607)100 times run just be more certain.